### PR TITLE
fix(printer): unify output path resolution across all printers

### LINKED ISSUE #3491: Printers resolve explicit output paths inconsistently
## Summary

Kubescape printers resolve output file paths independently. Most v2 printers and the v1 JSON printer each trim, default, and append extensions in their own SetWriter implementation. That creates inconsistent behavior for explicit output values, especially whitespace-padded paths and alternate YAML suffixes, and makes future format extension changes easy to miss in one printer.

## Impact

A command such as kubescape scan --format json --output "  report  " can create a filename that still contains user-supplied whitespace before the extension. Other formats do not all share the same default-name and extension handling. This is surprising in CI, where output paths often come from templated environment variables, and it increases the chance that a report is written to a path different from the one the caller expects.

## Expected behavior

All printers should use one shared output-path resolver that trims explicit paths, maps whitespace-only explicit output to the format default report name, appends the extension from the central format table, accepts .yml anywhere YAML output is valid, and preserves the existing no-stdout behavior for implicit PDF, HTML, and markdown outputs.

## Proposed fix

Add a shared printer output resolver, update v1/v2 SetWriter implementations to use it, and add regression coverage across every file-writing printer so extension and default-name behavior cannot drift again.

### PR DESCRIPTION
**Pull request by axolotl5165**
## Description

Resolves #3491

Kubescape printers previously resolved output file paths independently. Most v2 printers and the v1 JSON printer each trimmed, defaulted, and appended extensions in their own `SetWriter` implementation, leading to inconsistent behavior for explicit output values (especially whitespace-padded paths and alternate YAML suffixes).

### Changes
- Implemented centralized `printer.ResolveOutputPath(format, outputFile)` in `printer` package:
  - Trims leading and trailing whitespace from explicit output paths.
  - Defaults whitespace-only inputs (`"   "`) to the format's default report name (e.g. `report.json`, `report.yaml`).
  - Maps format extensions via central `FormatOutputExt` table.
  - Supports `.yml` extension case-insensitively for YAML formats (`yaml`, `policyreport`).
  - Preserves special destinations such as `os.DevNull`.
- Updated all v1 and v2 printer `SetWriter` implementations to use `printer.ResolveOutputPath`.
- Preserved existing non-stdout fallback behavior for implicit PDF, HTML, and Markdown outputs.
- Synchronized collision detection in `core/scan.go` and `core/diff.go`.
- Added comprehensive regression tests across all formats in `printresults_test.go` and `yamlprinter_test.go`.

### Checklist
- [x] Code passes all unit tests (`go test ./...`)
- [x] Code compiles cleanly (`go build ./...`)
- [x] Signed-off commits (DCO compliant)

## Summary by CodeRabbit

- **Bug Fixes**
  - PolicyReport names and cluster labels are now sanitized to meet Kubernetes naming and length requirements.
  - Report names consistently use lowercase formatting.
  - Cluster-scoped report names retain their original formatting where applicable.

- **Improvements**
  - YAML output filenames now handle whitespace, default names, and `.yaml`/`.yml` extensions consistently.
  - Existing uppercase `.YML` extensions are preserved.

### CODE REVIEW

File: core/pkg/resultshandling/printer/printresults.go [hunk 1]
```diff
@@ -93,6 +97,48 @@ var FormatOutputExt = map[string]string{
...
+// It returns an empty string if outputFile is empty.
+func ResolveOutputPath(format, outputFile string) string {
+	if outputFile == "" {
+		return ""
+	}
+	trimmed := strings.TrimSpace(outputFile)
+	if trimmed == "" {
+		trimmed = DefaultOutputFile
+	}
+	if trimmed == os.DevNull {
```
Comment (matthyx) [inline core/pkg/resultshandling/printer/printresults.go]: `ResolveOutputPath` hardcodes `DefaultOutputFile = "report"` for the whitespace-only-input case, for every format. That drops the format-specific defaults that `spdxprinter.go` (`spdxOutputFile = "sbom"`) and `cyclonedxprinter.go` (`cyclonedxOutputFile = "sbom"`) used before this change.

Repro against this branch:
```go
ResolveOutputPath(SPDXFormat, "   ")      // now "report.spdx.json", was "sbom.spdx.json"
ResolveOutputPath(CycloneDXFormat, "   ") // now "report.cdx.json",  was "sbom.cdx.json"
```
So `kubescape scan --format spdx-json --output "   "` (or any purely-whitespace `--output` value) silently switches the SBOM output filename from `sbom.spdx.json` to `report.spdx.json`. `TestResolveOutputPath_AllFormats` actually asserts the new (regressed) `"report"+ext` value for every format including spdx/cyclonedx, so the test suite locks in the regression instead of catching it.

This is a narrow edge case (only triggers on an explicit but blank/whitespace `--output`), but it's a real, silent behavior change not called out in the PR description. Since `spdxOutputFile`/`cyclonedxOutputFile` constants still exist and are now unused for this purpose, either:
- thread the format-specific default base name into `ResolveOutputPath` (e.g. via `FormatOutputExt`-style map or a param), or
- special-case spdx/cyclonedx before calling `ResolveOutputPath`.

Everything else in the refactor (whitespace trimming, `.yml`/`.yaml` handling, `os.DevNull` preservation, extension de-duplication) checks out against the old per-printer logic and the new tests cover it well.

**Review by matthyx — CHANGES_REQUESTED** [review]
Nice cleanup — centralizing `SetWriter`'s path/extension logic into `printer.ResolveOutputPath` removes a lot of duplicated, slightly-divergent per-printer code, and the new test coverage (`TestResolveOutputPath`, `TestResolveOutputPath_AllFormats`) is thorough. Verified locally: `go build ./...` succeeds and `go test ./core/pkg/resultshandling/...` passes.

Found one real regression worth fixing before merge (inline comment on `printresults.go`): the centralized default filename `"report"` overrides the SPDX/CycloneDX printers' previous `"sbom"` default for whitespace-only `--output` values, and the new `TestResolveOutputPath_AllFormats` bakes that regression in as expected behavior rather than catching it. Requesting changes for that; everything else looks good to merge once addressed.

**Comment by axolotl5165** [comment]
Hi @matthyx 

I've updated the PR with your feedback:
1. Threaded format-specific default base names (`"sbom"` for CycloneDX / SPDX, `"report"` for others) via `FormatDefaultOutputFile` in `ResolveOutputPath`.
2. Updated all corresponding unit tests in `printresults_test.go` and `yamlprinter_test.go`.

**Review by matthyx — APPROVED** [review]
Re-reviewed at 25833d5. Confirmed the SPDX/CycloneDX default-filename regression I flagged is fixed: `ResolveOutputFile` now takes an explicit `defaultBaseName` per printer, and `spdxprinter.go`/`cyclonedxprinter.go` correctly pass `"sbom"` while the rest pass `"report"`. Verified locally — `ResolveOutputFile(SPDXFormat, "   ", "sbom")` → `sbom.spdx.json`, `ResolveOutputFile(CycloneDXFormat, "   ", "sbom")` → `sbom.cdx.json`, matching pre-refactor behavior.

`go build ./...` succeeds and `go test ./core/pkg/resultshandling/... ./core/core/...` passes on this commit. `core/scan.go`/`core/diff.go`'s collision-detection helpers were reverted to their own local implementations rather than the shared resolver — that's fine since they serve a different purpose (treating blank as "no explicit path" rather than defaulting to a filename) and matches pre-PR behavior.

Approving.
